### ISSUE TITLE
JsonResult causes thread pool exhaustion via synchronous flushes

### ISSUE THREAD

**Issue by Tratcher** [aspnet/Mvc#8486]
_From @lorenzos70 on September 16, 2018 11:35_

Hello,

I have a .net 4.6.2 project (for my web apis) that is referencing, among other libraries, Microsoft.AspNetCore 2.1.3. I recently upgraded it from version 1.1. Since the upgrade I experienced the following issues on my production win 2016 server:
 
- my api logs often show the following messages: 
    1. Connection id ""xxxxxx"", Request id ""yyyyyy"": the connection was closed because the response was not read by the client at the specified minimum data rate
            
   2. the request timed out because it was not sent by the client at a minimum of 240 bytes/second".
           
   3. [Warning] Connection processing ended abnormally.
            Microsoft.AspNetCore.Connections.ConnectionAbortedException: The connection was aborted
            at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
           at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
           at System.IO.Pipelines.Pipe.GetReadAsyncResult()
           at System.IO.Pipelines.Pipe.DefaultPipeReader.GetResult(Int16 token)
           at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol. 
           <ProcessRequests>d__188`1.MoveNext()--- End of stack trace from previous location where 
          exception was thrown ---   at 
         System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)   at 
         System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task 
         task)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol. 
        <ProcessRequestsAsync>d__187`1.MoveNext()

- The apis become unresponsive (they are hosted on iis 10 on a windows 2016 server)
- IIS logs show, at the same time, 502 errors while trying to call the api endoints 
- I also noticed heavy peaks of the CPU.

Remarks: the web api load has not increased and there haven't been any other major changes, apart from the upgrade of the  Microsoft.AspNetCore that I mentioned above, so I am still trying to figure out what the problem could be. Any advice that could point me to the right direction?
Thanks
Lorenzo

_Copied from original issue: aspnet/KestrelHttpServer#2936_

**Comment by Tratcher** [comment]
_From @benaadams on September 16, 2018 13:32_

Upgrade to 2.1.4 and revisit https://github.com/aspnet/Announcements/issues/316

**Comment by Tratcher** [comment]
_From @lorenzos70 on September 16, 2018 13:57_

Hi Ben,

thanks for you reply: I tried but I realized that in my api project, which has been created using .NET 4.6 + .NET core (see first screenshot), the Nuget Packages manager doesn't show any release after the 2.1.3 (see second screenshot). 

I also noticed that the upgrade to 2.1.4 is instead available for "pure" .NET core projects. Do I need to install anything on my developer machine?

![cattura](https://user-images.githubusercontent.com/10574993/45597203-be2cb980-b9c8-11e8-9eb7-d3f521589d97.PNG)
![cattura vs](https://user-images.githubusercontent.com/10574993/45597205-c553c780-b9c8-11e8-873a-cfca20930c77.PNG)

**Comment by Tratcher** [comment]
_From @benaadams on September 16, 2018 14:0_

Ah, good point. Add a reference to (or upgrade) `System.IO.Pipelines` to 4.5.1 instead

**Comment by Tratcher** [comment]
_From @lorenzos70 on September 16, 2018 14:11_

Ok, thanks; I just added the System.IO.Pipelines refrerence (seems it wasn't installed before).

**Comment by Tratcher** [comment]
_From @benaadams on September 16, 2018 14:16_

> seems it wasn't installed before

It would have been brought in as a dependent reference from one of the other libraries.

Not sure if it will fix the data rate issue you are seeing; but it could fix the CPU spikes; and at a stretch they could be related. So see if it still happens.

**Comment by Tratcher** [comment]
_From @lorenzos70 on September 16, 2018 15:56_

Thanks again Ben; do you think that referencing the System.IO.Pipelines directly as I did (but still having the Microsoft.AspNetCore 2.1.3 referenced, which seems not using it) will make Kestrel to load  it anyway?

**Comment by Tratcher** [comment]
_From @benaadams on September 16, 2018 16:3_

Yes, Kestrel uses `System.IO.Pipelines` and adding a reference to `4.5.1` will cause it to use the patched version instead.

**Comment by Tratcher** [comment]
_From @lorenzos70 on September 17, 2018 8:31_

Unfortunately the problem occurred again even after the upgrade; I enabled the failed requests trace (see below) and I noticed that the response body's data buffers are read fine up to a certain point, and then the 502.3 error is raised by the  AspNetCoreModule (so it seems that is not the database query that is running slow, but the way the retrieved data is processed by the underlying layers). 
Moreover, the api logs also reveal the following error: Connection id ""xxxxyyyyy"" bad request data: "Reading the request body timed out due to data arriving too slowly. See MinRequestBodyDataRate".

![cattura](https://user-images.githubusercontent.com/10574993/45612756-4b711c00-ba64-11e8-9074-28153a781e85.PNG)

**Comment by Tratcher** [comment]
_From @lorenzos70 on September 18, 2018 11:49_

Hi ben,

just to let you know that I found this: [https://github.com/aspnet/KestrelHttpServer/issues/2104](url) which seems to describe exactly my situation. I indeed have some error handling methods that call something like this:
string  _result = ReadRequest(context).GetAwaiter().GetResult();

Could this cause the mentioned prtoblems? Adding something like this to main():

            var host = new WebHostBuilder()
                .UseKestrel()
                 .UseLibuv(options =>
                 {
                     options.ThreadCount = 1000;
                 })
                 .UseContentRoot(Directory.GetCurrentDirectory())
                .UseIISIntegration()
                .UseStartup<Startup>()
                .UseApplicationInsights()
                .Build();

could also help mitigate the problem?
Thanks again

**Comment by Tratcher** [comment]
_From @benaadams on September 18, 2018 13:20_

Wondering if the synchronous `Read` could deadlock the read for regular sockets; which will then get killed off due to its data rate dropping to 0 byte/s /cc @halter73

**Comment by Tratcher** [comment]
_From @halter73 on September 18, 2018 21:5_

```
    options.ThreadCount = 1000;
```

This will create 1000 managed threads outside the ThreadPool which is way too much. Generally, we find the best performance comes from having about `logical core count / 2` libuv threads. I don't recommend changing the libuv ThreadCount unless you've done extensive benchmarking proving it's better for all your scenarios.

As far as synchronous reads go, I'd advise against it since it can contribute to ThreadPool starvation. ThreadPool starvation (or even having 1000 libuv threads) is a possible cause request body rate timeouts if the starvation is so bad that read callbacks don't get scheduled in a timely manner.

I don't think a synchronous read could cause a deadlock in 2.1. There was a synchronous read deadlock in our 2.2 preview with HTTP/2 but that's in code that doesn't exist in 2.1. For HTTP/1.x, we always dispatch for reads unless you use a pubternal API to set the ApplicationSchedulingMode to Inline which you should *never* do if you're mixing sync and async I/O.

**Comment by Tratcher** [comment]
_From @noelstieglitz on September 18, 2018 21:8_

I started seeing a similar, sporadic issues yesterday in one Azure App Service and now am seeing it in a second:

` Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException: Reading the request body timed out due to data arriving too slowly. See MinRequestBodyDataRate.   at Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException.Throw(RequestRejectionReason reason)   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1MessageBody.PumpAsync()   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)   at System.IO.Pipelines.Pipe.GetReadAsyncResult()   at System.IO.Pipelines.Pipe.DefaultPipeReader.GetResult(Int16 token)   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.MessageBody.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.Read(Byte[] buffer, Int32 offset, Int32 count)   at System.IO.StreamReader.ReadBuffer()   at System.IO.StreamReader.ReadToEnd()   at ServiceStack.Text.JsonSerializer.DeserializeFromStream(Type type, Stream stream)   at ServiceStack.Serialization.JsonDataContractSerializer.DeserializeFromStream(Type type, Stream stream)   at ServiceStack.Host.Handlers.ServiceStackHandlerBase.CreateContentTypeRequest(IRequest httpReq, Type requestType, String contentType)`

I've only seen it happen on http `POST` operations.  I upgraded one of the App Service's code base to use `2.1.4` (and verified no transitive dependencies were pointing to the affected versions) and am still seeing the issue.  So far, it's not been reproducible locally, but it is sporadic, so I'll keep trying. 
As a workaround, I could configure kestrel to increase the `MinRequestBodyDataRate`, but that's obviously not a long term solution.

This issue is reminiscent of #2151 .

**Comment by Tratcher** [comment]
_From @halter73 on September 18, 2018 21:36_

@noelstieglitz How long were you running your app without seeing issues before seeing issues yesterday? Did anything else notable change regarding your app around the time you starting seeing the request body rate timeouts?

Also, what makes this reminiscent #2151 to you? One thing that helped us resolve #2151 was getting info from the reporter about how to reproduce the issue on our own hardware allowing us to investigate the issue more rapidly without losing time going back and forth getting more and more diagnostics from a third party.

One thing you can try that might help if the issue is that read callbacks are getting dispatched slowly due to ThreadPool starvation or similar is to change the `gracePeriod` component of the MinRequestBodyDataRate MinDataRate value. The default is 5 seconds, but it can be increased to whatever you want without changing the actual rate that's expected for large reads.

**Comment by Tratcher** [comment]
_From @lorenzos70 on September 19, 2018 14:2_

@halter73 I think I may have some more details about the root of my 502.3 api errors in my environment: I run a stress test on a specific asp.net core web api using the code suggested by @benaadams on a [previous issue](https://github.com/aspnet/KestrelHttpServer/issues/2104#issuecomment-337539008).

When the number of requests per minute hit 100 or more I noticed that the problems started to appear: api unresponsive, requests ending up in a 502.3 error, logs showing "Connection id ""0HLGU6OA7QP1G"", Request id ""0HLGU6OA7QP1G:00000001"": the connection was closed because the response was not read by the client at the specified minimum data rate", kestrel to be restarted in order to wake the server up again. 

The api returns a json array and it seems that the errors starts when the total size of the returned json objects (i.e. considering the number of requests by the size of the single response payload) exceeds a certain limit. This assumptions is due to the fact that, when stressing the very same api executing the very same code, with the only modification of returning an empty array instead of the calculated json object the test runs smoothly and no errors whatsoever show up. I am not very expert with the IIS and kestrel internals, so do you think that what I am saying makes sense? What is the parameter I could modify, apart from scaling up the servers?

**Comment by Tratcher** [comment]
It sounds like that throughput error is a symptom of your application having higher layer performance issues. If it can't handle 100 r/m then there is a significant bottleneck in the design. What to check next:
- How large is the response when you start having that problem? It's possible you are saturating your network bandwidth, though this should require responses in the GB range.
- Is the response very computationally expensive to generate? A profiler would be able to show you this. If your application is overtaxing the CPU then other parts of the system may start to fail. Making the generation more efficient or scaling up the application are your options here.
- Another version of this is that you may be generating and serializing the response inefficiently with regards to blocking threads. This can cause a lot of problems for other parts of the application trying to use the thread pool. If you attach a debugger and break then you would see a lot of threads blocked on wait handles.

**Comment by Tratcher** [comment]
_From @noelstieglitz on September 19, 2018 18:26_

@halter73 - thanks for the quick response. 

I did some more digging and found that the exception happened once on 8/21, 8/27, and 8/30.  Then a few times on the 14th and we've seen ~75 errors this week.  This could be due to increased traffic.  I'll need to write a fancy query to find out.

> How long were you running your app without seeing issues before seeing issues yesterday?

This app has been running for months.

> Did anything else notable change regarding your app around the time you starting seeing the request body rate timeouts?

We deployed a change that I initially suspected was related (based on the timing), but after rolling back that change I was still able to reproduce the error.  The only other piece that (I know) changed was the rollout of .net core `v2.1.4` to Azure App Services, which was applied to the specific app service instance on 9/14 (~the same day I first see the error in app insights~).  I didn't update the app's reference from `v2.1.2` to `v2.1.4` until yesterday (9/18), so I wouldn't expect that to be related?

> Also, what makes this reminiscent #2151 to you?

It's the same exception type (`BadHttpRequestException`), HTTP response (`400`) and hosted in Azure App Services (proxied through IIS first).  It happens sporadically.  The stack trace is different though and the other issue doesn't mention `MinRequestBodyDataRate`, so maybe not related at all.

>  it can be increased to whatever you want without changing the actual rate that's expected for large reads.

The response body is not large, in fact for one of the operations, the response body is empty.  The other response (including headers) is about 1,300 bytes.  Neither operations is CPU intensive.  They both make several asynchronous calls to other endpoints (some of which are parallel).  Typically the entire call takes something like a few hundred ms to maybe 2k ms.  When it fails, the request takes ~6k-7k ms.  The related code hasn't changed recently. Given these details, do you still think it makes sense to adjust the `MinRequestBodyDataRate::GracePeriod`?

> getting info from the reporter about how to reproduce the issue on our own hardware

Of course.  Short of creating a demo project, what would be the best way to do that?  Fiddler session?

**Comment by Tratcher** [comment]
_From @lorenzos70 on September 19, 2018 19:10_

@Tratcher thank you for the very precise analysis; I definetely believe it is the third scenario since the response body can hardly make up to GB in size and the response is generated quite efficiently in terms of CPU and resources; besides when I run this:

        [HttpGet]
        public JsonResult TesterApiMethod()
        {
            IEnumerable<myDTO> response = GetData();
            return Json("OK");
        }
everything works like a charm (I tried up to 350 requests/min), 
while if I run this:

        [HttpGet]
        public JsonResult TesterApiMethod()
        {
            IEnumerable<myDTO> response = GetData();
            return Json(response);
        }
  Kestrel stops responding and processing the requests even around 150 requests/minute. myDTO is a quite standard DTO object.

**Comment by Tratcher** [comment]
Is that IEnumerable lazy? If you change it to `GetData().ToList();` in your second sample does that change the performance?

What does the client report the final response size as?

It looks like JsonResult is writing synchronously, which should only be a problem if the sizes get very large. Adding response buffering may help. Here's a prototype for that you could try:
https://github.com/aspnet/BasicMiddleware/blob/2a731baaa3a4f6c91abad0bd68b1ea01f025d10a/src/Microsoft.AspNetCore.Buffering/ResponseBufferingMiddlewareExtensions.cs

**Comment by Tratcher** [comment]
_From @lorenzos70 on September 19, 2018 22:30_

 @Tratcher the IEnumerable is of type List<T>, so I haven't gained any performance benefit; as far as the response size is concerned, it is roughly 90k, hence nothing too critical. I am then still struggling to find why the threads get blocked this way during serialization.

**Comment by Tratcher** [comment]
_From @lorenzos70 on September 20, 2018 11:9_

@Tratcher I enabled response buffering and it dramatically impacted the performance: so far I stress tested with 1000 req/min without any problem (I am confident it can handle much more now), while when disabling response buffering the server gets stuck even with only 150 req/min. Does it make sense with my limited response size? I am still very surprised for this behavior.

**Comment by Tratcher** [comment]
That's not a good sign. Can you attach the debugger when it gets stuck, break all, and look at the current thread call stacks? Are there a lot of threads blocked on wait handles? What are the common call stacks?

**Comment by Tratcher** [comment]
@sebastienros is [this](https://github.com/aspnet/KestrelHttpServer/issues/2936#issuecomment-422923588) covered in any of the perf benchmarks?

**Comment by Tratcher** [comment]
_From @lorenzos70 on September 20, 2018 15:50_

@Tratcher I managed to dump the hanging process and run a report (available as a .mht file at https://drive.google.com/open?id=1oFEvZl-RmDJbGnen_z4lnCSMCg7WTaJc), which confirms the amount of threads blocked on waits.

**Comment by Tratcher** [comment]
_From @benaadams on September 20, 2018 16:5_

Trimming some noise from the thread stacks:
```
System.Threading.Monitor.ObjWait(Boolean, Int32, System.Object) 
mscorlib_ni!System.Threading.Monitor.Wait(System.Object, Int32, Boolean)+17 
mscorlib_ni!System.Threading.ManualResetEventSlim.Wait(Int32, System.Threading.CancellationToken)+226 
mscorlib_ni!System.Threading.Tasks.Task.SpinThenBlockingWait(Int32, System.Threading.CancellationToken)+9c 
mscorlib_ni!System.Threading.Tasks.Task.InternalWait(Int32, System.Threading.CancellationToken)+155 
Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpResponseStream.Write(Byte[], Int32, Int32)+4d 
Microsoft.AspNetCore.WebUtilities.HttpResponseStreamWriter.FlushInternal(Boolean)+51 
Microsoft.AspNetCore.WebUtilities.HttpResponseStreamWriter.Write(Char)+20 
Newtonsoft.Json.JsonTextWriter.WriteEnd(Newtonsoft.Json.JsonToken)+29 
Newtonsoft.Json.JsonWriter.AutoCompleteClose(Newtonsoft.Json.JsonContainerType)+69 
Newtonsoft.Json.JsonWriter.WriteEndObject()+b 
Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(Newtonsoft.Json.JsonWriter, System.Object, Newtonsoft.Json.Serialization.JsonObjectContract, Newtonsoft.Json.Serialization.JsonProperty, Newtonsoft.Json.Serialization.JsonContainerContract, Newtonsoft.Json.Serialization.JsonProperty)+2db 
...
Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(Newtonsoft.Json.JsonWriter, System.Object, Newtonsoft.Json.Serialization.JsonObjectContract, Newtonsoft.Json.Serialization.JsonProperty, Newtonsoft.Json.Serialization.JsonContainerContract, Newtonsoft.Json.Serialization.JsonProperty)+f9 
...
Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize(Newtonsoft.Json.JsonWriter, System.Object, System.Type)+a9 
Newtonsoft.Json.JsonSerializer.SerializeInternal(Newtonsoft.Json.JsonWriter, System.Object, System.Type)+311 
Microsoft.AspNetCore.Mvc.Formatters.JsonOutputFormatter.WriteObject(System.IO.TextWriter, System.Object)+47 

Microsoft.AspNetCore.Mvc.Formatters.JsonOutputFormatter+<WriteResponseBodyAsync>d__11.MoveNext()+b8 
Microsoft.AspNetCore.Mvc.Formatters.JsonOutputFormatter.WriteResponseBodyAsync(Microsoft.AspNetCore.Mvc.Formatters.OutputFormatterWriteContext, System.Text.Encoding)+73 
Microsoft.AspNetCore.Mvc.Formatters.TextOutputFormatter.WriteAsync(Microsoft.AspNetCore.Mvc.Formatters.OutputFormatterWriteContext)+11e 
Microsoft.AspNetCore.Mvc.Infrastructure.ObjectResultExecutor.ExecuteAsync(Microsoft.AspNetCore.Mvc.ActionContext, Microsoft.AspNetCore.Mvc.ObjectResult)+1b6 
Microsoft.AspNetCore.Mvc.ObjectResult.ExecuteResultAsync(Microsoft.AspNetCore.Mvc.ActionContext)+2b 
Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker+<InvokeResultAsync>d__20.MoveNext()+7a 
...
```

Looks like 
`Microsoft.AspNetCore.Mvc.Formatters.JsonOutputFormatter.WriteResponseBodyAsync`
via
`Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject`
ending up calling 
`Microsoft.AspNetCore.WebUtilities.HttpResponseStreamWriter.Write(Char)`
which then hits
`System.Threading.Tasks.Task.InternalWait`

**Comment by Tratcher** [comment]
@davidfowl @rynowak @JamesNK

**Comment by rynowak** [comment]
Yep. We should add a call to proactively flush the writer.
**GOOD** https://github.com/aspnet/Mvc/blob/release/2.2/src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonOutputFormatter.cs#L153
**BAD** https://github.com/aspnet/Mvc/blob/release/2.2/src/Microsoft.AspNetCore.Mvc.Formatters.Json/Internal/JsonResultExecutor.cs

**Comment by sebastienros** [comment]
Not sure I am repeating this, long thread, but @lorenzos70 can you try to remove these `return JSon(obj)` calls and fallback to the output formatter by doing this:

```csharp
    [HttpGet]
    public ActionResult<myDTO> TesterApiMethod()
    {
        var response = GetData();

        if (response.Any())
        {
            return response;
        }
        
        return NoContent();
    }
```

**Comment by lorenzos70** [comment]
@sebastienros I tried your approach but had the impression that the same serializers are applied and that dump produces the identical stack trace with the same waits, but maybe I am missing some additional setting (I also removed the Produces("application/json") attribute).

**Comment by pranavkm** [comment]
@lorenzos70 another workaround outside using of using the response buffering middleware is to increase the buffer size used by `JsonResult`, and the JsonOutputFormatter (used in the code path @sebastienros pointed to). There isn't a setting to tweak it, but writing a custom factory that specifies the size is pretty trivial. Add a writer factory to your application and tweak the DefaultBufferSize to a larger value (you could start by doubling the current default):

```C#
using System;
using System.Buffers;
using System.IO;
using System.Text;
using Microsoft.AspNetCore.Mvc.Infrastructure;
using Microsoft.AspNetCore.WebUtilities;

namespace Microsoft.AspNetCore.Mvc
{
    public class CustomHttpResponseStreamWriterFactory : IHttpResponseStreamWriterFactory
    {
        public static readonly int DefaultBufferSize = 16 * 1024;

        private readonly ArrayPool<byte> _bytePool;
        private readonly ArrayPool<char> _charPool;

        public CustomHttpResponseStreamWriterFactory(ArrayPool<byte> bytePool, ArrayPool<char> charPool)
        {
            _bytePool = bytePool;
            _charPool = charPool;
        }

        public TextWriter CreateWriter(Stream stream, Encoding encoding)
        {
            return new HttpResponseStreamWriter(stream, encoding, DefaultBufferSize, _bytePool, _charPool);
        }
    }
}
```

2) Register the factory in DI as part of your startup:

```C#
public void ConfigureServices(IServiceCollection services)
{
    services.AddSingleton<IHttpResponseStreamWriterFactory, CustomHttpResponseStreamWriterFactory>();
}
```
3. Use output formatter (i.e. returning the object or returning `ObjectResult` instead of `JsonResult`) until 2.2 becomes available.

**Comment by lorenzos70** [comment]
@pranavkm thanks for the snippets, very useful; by the way, I see that on October 2nd version 2.1.3 has been published but I assume that the patch you are working on will be available in version 2.2, is that correct?

**Comment by pranavkm** [comment]
> but I assume that the patch you are working on will be available in version 2.2, is that correct?

Yes. The fix is in our 2.2 release.
